api/classes/publisher.js

```diff
@@ -79,72 +115,236 @@ function deleteFile(file, root) {
+      if (oldPublishedProject) {
+        return oldPublishedProject.set({
+          title: project.get('title'),
+          tags: project.get('tags'),
+          description: project.get('description')
+        }).save();
+      }
+      return PublishedProject.forge({
+        title: project.get('title'),
+        tags: project.get('tags'),
+        description: project.get('description')
+      }).save();
+    })
+    .then(function(p) {
+      publishedProject = p;
+    });
   })
-  .catch(function(e) {
-    return Promise.reject(e);
-  });
-}
+  .then(function() {
+    return File.query({
+      where: {
+        project_id: project.get('id')
+      }
+    }).fetchAll()
+    .then(function(f) {
+      if (f.length === 0) {
+        throw 'No files to be published!';
+      }
 
-function getUser(project) {
-  return project.user().query({}).fetch();
-}
+      files = f;
+    });
+  })
+  .then(function() {
+    return PublishedFiles.query({
+      where: {
+        published_id: publishedProject.get('id')
+      }
+    }).fetchAll()
+    .then(function(f) {
+      if (f.length === 0) {
+        return;
+      }
+      publishedFiles = f;
+    });
+  })
+  .then(function() {
+    // We want to organize 'file' objects by id, and
+    // 'publishedFile' objects by the 'file' objects they reference
+    files.forEach(function(file) {
+      projectFiles[file.get('id')] = file;
+    });
+    var filePrimaryKeys = Object.keys(projectFiles);
 
-exports.publish = function(project) {
-  return getUser(project)
-    .then(function(user) {
-      return modifyFiles(uploadFile, project, user);
-    })
-    .then(function(url) {
-      return project.set({
-        publish_url: url
+    // Nothing published yet?
+    if (!publishedFiles) {
+      return filePrimaryKeys.forEach(function(id) {
+        return toBeCreated.push(projectFiles[id]);
+      });
+    }
+
+    // We collect references to each model that needs to
+    // have its 'publishedFile' created or destroyed (respectively)
+    publishedFiles.forEach(function(publishedFile) {
+      if (publishedFile.get('file_id') === null) {
+        toBeDeleted.push(publishedFile);
+        return;
+      }
+      publishedProjectFiles[publishedFile.get('file_id')] = publishedFile;
+    });
+    var publishedFileForeignKeys = Object.keys(publishedProjectFiles);
+
+    filePrimaryKeys.forEach(function(primaryKey) {
+      if (publishedFileForeignKeys.indexOf(primaryKey) === -1) {
+        toBeCreated.push(projectFiles[primaryKey]);
+      }
+    });
+    publishedFileForeignKeys.forEach(function(foreignKey) {
+      // If the reference exists in a 'publishedFile' object as
+      // a foreign key, and in a 'file' object as a primary key,
+      // we will be forcing an update in case the file changed before
+      // republishing.
+      if (filePrimaryKeys.indexOf(foreignKey) !== -1) {
+        toBeUpdated.push(publishedProjectFiles[foreignKey]);
+      }
+    });
+  })
+  .then(function() {
```

Comment:
I think it's more clear this way when you have them as separate functions for create, update and delete